### PR TITLE
ch4/ofi: Avoid using generic object free

### PR DESCRIPTION
**Pull request by raffenet**
Request alloc/free functionality is provided by the upper layer.

### CODE REVIEW

**Comment by raffenet** [comment]
test:jenkins/ch4/ofi

**Comment by raffenet** [comment]
There was an issue with the win req complete function. Made some changes to be squashed if the testing passes.
test:jenkins/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 1]
```diff
@@ -302,17 +302,11 @@ MPL_STATIC_INLINE_PREFIX void MPIDI_OFI_win_datatype_unmap(MPIDI_OFI_win_datatyp
...
-        MPIDI_OFI_win_datatype_unmap(&req->noncontig->origin_dt);
-        MPIDI_OFI_win_datatype_unmap(&req->noncontig->result_dt);
-        MPL_free(req->noncontig);
-        MPIR_Request_free((req));
-    }
+    MPIDI_OFI_win_datatype_unmap(&req->noncontig->target_dt);
+    MPIDI_OFI_win_datatype_unmap(&req->noncontig->origin_dt);
+    MPIDI_OFI_win_datatype_unmap(&req->noncontig->result_dt);
+    MPL_free(req->noncontig);
+    MPIR_Request_free(req);
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: Is it safe to do this? i.e. there's no cases where `req` has refcount more than one? I guess not, but just wanted to confirm.

Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: Well, directly above that we assert that the last reference has been released, so my best guess was that this was safe. Testing seems to confirm.

Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: Oh, wait. I see what you mean.

**Comment by raffenet** [comment]
I only bit off a small part in this new version. The other instances of using `MPIR_Handle_obj_free` are using special internal request types that I want to think about more.
test:jenkins/ch4/ofi

**Comment by raffenet** [comment]
test:jenkins/ch4/ofi

**Pull request merged**